pkg/cluster/components/daemonset.go

```diff
@@ -0,0 +1,178 @@
+package components
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	"k8s.io/kubernetes/pkg/apis/extensions/v1beta1"
+	"k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_3"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/deployment"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+// DaemonSetUpdater can provide rolling updates on
+// the given DaemonSet.
+type DaemonSetUpdater struct {
+	// name of the DaemonSet object.
+	name string
+	// client is an API Server client.
+	client clientset.Interface
+	// podStore is backed by an informer, contains list of Pods
+	// that belong to the DaemonSet.
+	podStore cache.Store
+}
+
+func NewDaemonSetUpdater(client clientset.Interface, ds *v1beta1.DaemonSet) (*DaemonSetUpdater, error) {
+	var spec extensions.DaemonSetSpec
+	err := v1beta1.Convert_v1beta1_DaemonSetSpec_To_extensions_DaemonSetSpec(&ds.Spec, &spec, nil)
+	if err != nil {
+		return nil, err
+	}
+	selector, err := unversioned.LabelSelectorAsSelector(spec.Selector)
+	if err != nil {
+		return nil, err
+	}
+	lo := api.ListOptions{LabelSelector: selector}
+	podStore, podController := framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(_ api.ListOptions) (runtime.Object, error) {
+				return client.Core().Pods(ds.Namespace).List(lo)
+			},
+			WatchFunc: func(_ api.ListOptions) (watch.Interface, error) {
+				return client.Core().Pods(ds.Namespace).Watch(lo)
+			},
+		},
+		&v1.Pod{},
+		30*time.Minute,
+		framework.ResourceEventHandlerFuncs{},
+	)
+	go podController.Run(wait.NeverStop)
+	return &DaemonSetUpdater{
+		name:     ds.Name,
```

Comment:
We should also track namespace rather than assuming 'kube-system' (and we have that info available here)